agent-bootstrap/src/main/java/io/opentelemetry/auto/bootstrap/instrumentation/decorator/HttpServerTracer.java

```diff
@@ -195,6 +201,63 @@ protected void onRequest(final Span span, final REQUEST request) {
     // TODO set resource name from URL.
   }
 
+  protected void onConnectionAndRequest(Span span, CONNECTION connection, REQUEST request) {
+    String flavor = flavor(connection, request);
+    if (flavor != null) {
+      SemanticAttributes.HTTP_FLAVOR.set(span, flavor);
+    }
+    SemanticAttributes.HTTP_CLIENT_IP.set(span, clientIP(connection, request));
+  }
+
+  protected String clientIP(CONNECTION connection, REQUEST request) {
+    // try Forwarded
+    String forwarded = requestHeader(request, "Forwarded");
+    if (forwarded != null) {
+      forwarded = extractForwardedFor(forwarded);
+      if (forwarded != null) {
+        return forwarded;
+      }
+    }
+
+    // try X-Forwarded-For
+    forwarded = requestHeader(request, "X-Forwarded-For");
+    if (forwarded != null) {
+      // may be split by ,
+      int endIndex = forwarded.indexOf(',');
+      if (endIndex > 0) {
+        forwarded = forwarded.substring(0, endIndex);
+      }
+      if (!forwarded.isEmpty()) {
+        return forwarded;
+      }
+    }
+
+    // fallback to peer IP if there are no proxy headers
+    return peerHostIP(connection);
+  }
+
+  // VisibleForTesting
+  static String extractForwardedFor(String forwarded) {
+    int start = forwarded.toLowerCase().indexOf("for=");
```

Comment:
Since the spec seems to use lowercase, and it's not a commonly used header anyways, I probably wouldn't be so permissive and just not do the `toLowerCase` but it's ok too. From what I can tell, we consider it unimportant enough to not have unit tests :P

Reply:
the mozilla spec says it can be any case, so I just stuck that in there. I've never seen it used in practice but it's apparently the standard.

Reply:
there's an uppercase example in the RFC https://tools.ietf.org/html/rfc7239#section-4

Reply:
Ah thanks I couldn't find it, in that case can we add one test case for it?

Reply:
sure, added